src/utils/cache.ts

```diff
@@ -0,0 +1,45 @@
+import { stat } from './fs';
+import { HookCache, Plugin } from './rollup/types';
+
+export function cacheResolveIdHook(cache: HookCache, plugin: Plugin) {
+	return function cachedResolveId(id: string) {
+		const key = `resolve|${plugin.name || '(anonymous plugin)'}|${plugin.cacheKey}|${id}`;
```

Comment:
My concern with resolver caching is that we aren't tracking invalidation on the stat paths of the resolver itself, but instead relying on any resolve plugin to change its cacheKey somehow (likely a time interval?) based on that information, which might not be obvious.

One of the goals for the cache is to be able to by-default persist to disk out-of-the-box and I think that we should try and aim for this in the caching work. As a result we should be more cautious than that I think.

So my first thought on this would be to say we shouldn't cache the resolver for now, and can come back to this in due course (and since you say it is only 15% of the problem...), but let me know if you have further thoughts here.

Reply:
I agree entirely with these constraints. Here's some ideas:

Giving keys a TTL is easy enough:

```js
const key = `resolve|${plugin.name || '(anonymous plugin)'}|${plugin.cacheKey}|${id}|${~~(Date.now()/60000)}`
```
Of course a TTL isn't really the best for this - because the cache could be invalidated before the TTL is up, or could still be valid past the cache's timeout.

Another thing we could do, is read from the cache and compare it to what was known about the file at the time - e.g. `stat()` calling. This way we capture the external state (the file). Reading ahead I see there is a constraint on the browser - I'll address that below.

```js
const key = `resolve|${plugin.name || '(anonymous plugin)'}|${plugin.cacheKey}|${id}`
if (cache[key] && cache[key].path) {
  // Now we have a cache key we can double check to ensure the mtime/size are the same as what is on disk
  // and only return if they match
  const { mtimeMs, size } = await stat(cache[key].path).catch(() => ({ mtimeMs: 0, size: 0 }))
  if (mtimeMs === cache[key].mtimeMs && size === cache[key].size) {
    return Promise.resolve(cache[key])
  }
}
const path = await plugin.resolveId.call(this, id)
// Now we have the path we can stat it to get mtime and size for it
const { mtimeMs, size } = await stat(cache[key].path).catch(() => ({ mtimeMs: 0, size: 0 }))
cache[key] = { path, mtime, size }
return path
```

Reply:
The issue with statting the file directly, is edge cases like having a package.json where there is no main, and then a main gets added. Statting the index.js file won't tell you that the package.json file was changed to add a main. And if you stat the package.json file, and the next edge case and the next we're doing all the work of the resolver when it comes down to it.

So I tend to think the best cache for a resolver is a TTL-based one, and that the implementation for this should probably still be further downstream from Rollup itself.

Would you consider a PR to the rollup-plugin-node-resolve project here to add a resolver with this kind of functionality? We would need a way to add the plugin state to the Rollup cache object, which perhaps may be something we can open up from the Rollup side with something like `this.setCache('key', value)` and `this.getCache('key', value)` for plugins or similar. Let me know what you think along those lines?